Revamp for GNOME 45

adhadse opened the pull request:
Resolves #206 

- [X] Revamp as per https://gjs.guide/extensions/upgrading/gnome-shell-45.html#port-extensions-to-gnome-shell-45 and https://github.com/swsnr/gnome-shell-extension-typescript-template/tree/e8116afac1ad690d300985183238b51044445a3d
- [ ] Change `scripts/transpile.ts` to accommodate the above point
       
       file:///mnt/Data/vscode_repo/gnome-gesture-improvements/build/scripts/transpile.js:174
            throw new Error(`Unknown import statement '${node.getFullText()}'`);
                  ^

       Error: Unknown import statement ' 
       import { Extension } from 'resource:///org/gnome/shell/extensions/extension';'

adhadse committed:
import as per GJS ESM documentation

adhadseKavida commented:
I don't really have a background in TS/GJS and would appreciate someone's help on this for review and further refactoring.

smedir commented on metadata.json:
```diff
@@ -4,10 +4,11 @@
 		"42",
 		"43",
-		"44"
+		"44",
```
import will not work on 44

adhadse replied:
Oh, yes removing it.

smedir commented:
> Resolves #206
> 
> * [x]  Revamp as per https://gjs.guide/extensions/upgrading/gnome-shell-45.html#port-extensions-to-gnome-shell-45 and https://github.com/swsnr/gnome-shell-extension-typescript-template/tree/e8116afac1ad690d300985183238b51044445a3d
> * [ ]  Change `scripts/transpile.ts` to accommodate the above point
>   ```
>    file:///mnt/Data/vscode_repo/gnome-gesture-improvements/build/scripts/transpile.js:174
>         throw new Error(`Unknown import statement '${node.getFullText()}'`);
>               ^
>   
>    Error: Unknown import statement ' 
>    import { Extension } from 'resource:///org/gnome/shell/extensions/extension';'
>   ```

I was refactoring this as well, :+1: , had just started. You are way ahead. All yours.

smedir commented on extension/extension.ts:
```diff
@@ -13,16 +13,17 @@ import { SnapWindowExtension } from './src/snapWindow';
 import * as VKeyboard from './src/utils/keyboard';
 
-const ExtensionUtils = imports.misc.extensionUtils;
+const extensionObject = Extension.lookupByUUID('gestureImprovements@gestures');
```
This is not necessary

smedir commented on extension/extension.ts:
```diff
@@ -13,16 +13,17 @@ import { SnapWindowExtension } from './src/snapWindow';
 	private _addReloadDelayFor: AllSettingsKeys[];
 
 	constructor() {
+		super(extensionObject.metadata);
```
constructor(metadata: ExtensionMetadata) {
                  super(metadata)

smedir commented on extension/extension.ts:
```diff
@@ -13,16 +13,17 @@ import { SnapWindowExtension } from './src/snapWindow';
+const extensionObject = Extension.lookupByUUID('gestureImprovements@gestures');
 
-class Extension {
+class GNOMEGestureImprovementsExtension extends Extension {
```
export default class GnomeGestureImprovements extends  Extension {

line 156 is not necessary??

smedir commented on extension/extension.ts:
```diff
@@ -1,5 +1,5 @@
 import GLib from '@gi-types/glib2';
-import { imports } from 'gnome-shell';
+import { Extension } from 'resource:///org/gnome/shell/extensions/extension';
```
( Extension, ExtensionMetadata} from ....

adhadse committed:
bug fixes

- remove older version of gnome from supported list
- Add `ISubExtension` and `IExtension` to `global.d.ts`.

adhadseKavida commented:
Maybe I can learn from this [PR](https://github.com/oae/gnome-shell-pano/pull/222) for pano.

adhadse commented:
Maybe I can learn from this [PR](https://github.com/oae/gnome-shell-pano/pull/222) for pano (clipboard manager)

MaddyGuthridge commented:
Is there any news on this? This extension is currently keeping me on Fedora 38, which goes EOL in about 3 months' time. If there's anything I can do to get the extension updated, please let me know!

sidevesh commented:
It would be great if this branch can be merged in,
in the meantime I have pushed a release on https://github.com/sidevesh/gnome-gesture-improvements--transpiled which contains the transpiled version with fixes that a bunch of folks have done to get it working on Gnome 46.
